### PR TITLE
Enable grid outer persistent scheduling

### PR DESCRIPTION
**Pull request by naoyam**
(Extracted from #1772)
(Stacked on top of #2416)

Extends the normalization scheduler with grid persistence. The scheduling approach is mostly the same as the one I manually did here: https://github.com/csarofeen/pytorch/blob/devel/third_party/nvfuser/test/test_gpu_outer_reduction.cpp#L379. 

There are still many performance limitations, and in this PR I tried to minimize regressions compare to the current TOT even when that means missing perf gains in some other cases. Will address those limitations later.

Some major limitations:

- The non-welford grouped grid reduction kernel is slow. It's even slower than the welford kernel as it's specifically optimized for outer reductions, whereas the non-welford kernel is a generic template implementation.
- In fusions like half-precision batchnorm with float weights, the vectorization factor is limited to 4 due to the type of the weights inputs
- Iteration domains of size non power-of-two need better tuning. For example, the TIMM batchnorm benchmarks have iteration domains of size 200, whereas the ResNet batchnorms are easier to work with as their sizes are, e.g., 256, 512, etc.
- Using shared memory as well as L2 for additional cache space

### A100
Bandwidth and speedup curves of all the benchmarks. The left axis shows the bandwidth of each benchmark with the TOT devel and this branch (red and blue dots). The right axis shows the speed-up factor of each benchmark. The benchmark plots are sorted by the bandwidth of the devel branch. Note that the number of benchmarks that are scheduled as grid persistence is still small, so most of the dots are scattered around 1.0.

As I mentioned above, I tried to minimize performance regressions, and these graphs show there's indeed no significant regression. I saw even more benchmarks showing positive speedups with different heuristics configurations, but the current heuristics are the best I found so far without causing any major regression, at least on A100. (There are some regressions on Titan RTX as shown below.)
 
image

Speedup histogram. Note the Y axis is log scale.
image

The largest speedup is with this benchmark:
<html>
<body>

 name | bw_devel | bw_new | bw_new/bw_devel
-- | -- | -- | --
NvFuserScheduler_BatchNorm_nhwc_fp32___GRAPH/NvFuserScheduler_BatchNorm_nhwc_fp32/512/64/8/manual_time | 2.836540e+11 | 6.599990e+11 | 2.326775

</body>
</html>

It's still 660 GB/s, so likely there's still a lot of room to improve.

This one looks quite good now.

<html>
<body>

name | bw_devel | bw_new | bw_new/bw_devel
-- | -- | -- | --
NvFuserScheduler_ResNet_BatchNorm_nhwc_fp16___GRAPH/NvFuserScheduler_ResNet_BatchNorm_nhwc_fp16/256/2048/7/manual_time | 6.192360e+11 | 9.104440e+11 | 1.47027

</body>
</html>

### Titan RTX

The results on Titan RTX look more mixed compared to those on A100. There are some benchmarks exhibiting more than 20% regression, although they are quite small problems. It turned it's a little more difficult to avoid regressions on Titan RTX than on A100, and I even needed to add extra check for the 7.5 compute architecture. Quick tests with manual scheduling seems to indicate lack of full vectorization is one of the reasons. Will revisit again.

image

image

### V100

There's one case that got around 3.5x speedup, though it could be just a random noise given its raw bandwidth is just a few GB/s. Otherwise, the speedup factors look similar to those on A100.

image

image

### CODE REVIEW

File: third_party/nvfuser/csrc/scheduler/reduction_utils.cpp [hunk 1]
```diff
@@ -187,31 +219,67 @@ TensorView* scheduleReductionTV(
...
       }
     }
   }
 
-  return sortAndRFactor(reduction_tv);
+  auto reduction_rf_tv = sortAndRFactor(reduction_tv);
+
+  // In the case of outer grid persistence, make sure the vectorized
+  // domain placed at the innermost position.
+  // TODO: Why isn't this the case by default?
```
Comment (naoyam) [inline third_party/nvfuser/csrc/scheduler/reduction_utils.cpp:261]: Why doesn't `sortAndRFactor` move vectorized iteration domains to the innermost position?

File: third_party/nvfuser/csrc/scheduler/registry.cpp [hunk 2]
```diff
@@ -1839,6 +1842,202 @@ class PersistentKernelScheduler : public SchedulerEntry {
...
+        std::any_of(
+            reduction_tvs.begin(),
+            reduction_tvs.end(),
+            [](TensorView* reduction_tv) {
+              return !reduction_tv->definition()->isA<WelfordOp>();
+            })) {
+      scheduler_debug_utils::canScheduleRejectReason(
+          ScheduleHeuristic::Persistent, "non-Welford not enabled yet");
+      return false;
+    }
```
Comment (naoyam) [inline third_party/nvfuser/csrc/scheduler/registry.cpp:2021]: I'll remove this constraint once I optimize the non-welford grouped reduction kernel as I did the welford kernel.

File: third_party/nvfuser/csrc/scheduler/registry.cpp [hunk 3]
```diff
@@ -1839,6 +1842,202 @@ class PersistentKernelScheduler : public SchedulerEntry {
...
+              cross_grid_params->launch_params.gdimx()) !=
+         0) &&
+        device_prop->major == 7 && device_prop->minor == 5) {
+      scheduler_debug_utils::canScheduleRejectReason(
+          ScheduleHeuristic::Persistent, "iteration not evenly divided");
+      return false;
+    }
+
+    return true;
+  }
```
Comment (naoyam) [inline third_party/nvfuser/csrc/scheduler/registry.cpp]: Saw some regressions with the TIMM batcnorm benchmarks on Titan RTX. They have iteration domains of size, e.g., 200 and 368, which result in non-even distribution of work among threads and blocks. These cases are also slow on A100, but still the persistent version is better than the current segmented kernels.

Will revisit later.

**Comment by naoyam** [comment]
The V100 results actually showed non-negligible regressions. I didn't notice due to the outlier making. 

Added the same bail-out condition as Titan RTX: #26d1588

Updated performance results. The outlier result is excluded.

image

image

There's still a small number of regressions of about 20%. Part of the reasons seems to be due to this register usage estimate is too conservative: https://github.com/csarofeen/pytorch/blob/26d15886b77c5fbe66b7fc23c3c4065bc7ea9b42/third_party/nvfuser/csrc/scheduler/normalization_utils.cpp#L161. Will address later.

File: third_party/nvfuser/csrc/scheduler/normalization_utils.h [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+namespace cuda {
+namespace normalization_scheduler_utils {
+
+//! Utility class to iterate candidates of launch configurations in a
+//! preferred order. The iteration order is defined as:
+//!
+//!   for bdimx in all valid bdimx in an decreasing order
+//!     for gdimy in valid gdimy values in an increasing order
+//!
+//! Each of bdimx and gdimy determines gdimy and gdimx, respecitively,
```
Comment (csarofeen) [inline third_party/nvfuser/csrc/scheduler/normalization_utils.h]: Should this be Each of bdimx and *bdimy ?

File: third_party/nvfuser/csrc/scheduler/reduction_utils.cpp [hunk 5]
```diff
@@ -89,7 +98,31 @@ TensorView* scheduleReductionTV(
...
 
-  if (rparams.persistent_kernel) {
+  if (is_outer_grid_persistence) {
+    const auto reduction_axis = inner_reduce_axis;
+    TORCH_INTERNAL_ASSERT(rparams.static_bdimy, "blockDim.y must be static");
+    inner_parallel_static(
+        reduction_axis,
+        rparams.block_dim_inner_reduction,
+        rparams.lparams.bdimy());
+    reduction_tv->split(
```
Comment (csarofeen) [inline third_party/nvfuser/csrc/scheduler/reduction_utils.cpp:108]: Isn't this the point of `inner_parallel_static`?

File: third_party/nvfuser/csrc/scheduler/reduction_utils.cpp [hunk 6]
```diff
@@ -187,31 +219,67 @@ TensorView* scheduleReductionTV(
     // [Grid Split, unswitch, unroll, thread dim, vectorize]
 
     if (rparams.vectorize_iter_dom) {
-      vectorize(iter_axis, rparams.unroll_factor_iter_dom);
+      if (is_outer_grid_persistence) {
+        vectorize(iter_axis, rparams.unroll_factor_iter_dom);
+      } else {
+        vectorize(iter_axis, rparams.unroll_factor_iter_dom);
+      }
```
Comment (csarofeen) [inline third_party/nvfuser/csrc/scheduler/reduction_utils.cpp]: Cleanup?

File: third_party/nvfuser/csrc/scheduler/reduction_utils.cpp (hunk 1 above)
Comment (csarofeen) [reply third_party/nvfuser/csrc/scheduler/reduction_utils.cpp:261]: `sortAndRFactor` order was primarily just empirical. If al the C++ tests and benchmarks run with vectorized dimensions fully innermost then it's fine to change the order. I believe it could be easier to deal with now as I think it was mostly fighting with compute at.

File: third_party/nvfuser/csrc/scheduler/reduction_utils.cpp [hunk 7]
```diff
@@ -239,6 +307,44 @@ std::vector<int> addBackBroadcasts(
   return axes;
 }
 
+// Check if a reduction is effectively an allreduce.
+bool isAllreduce(TensorView* reduction_tv) {
```
Comment (csarofeen) [inline third_party/nvfuser/csrc/scheduler/reduction_utils.cpp]: Nit: isGridAllReduce

File: third_party/nvfuser/csrc/scheduler/registry.cpp [hunk 8]
```diff
@@ -1839,6 +1842,202 @@ class PersistentKernelScheduler : public SchedulerEntry {
...
+
+    const int64_t device_multiprocessor_count =
+        (int64_t)device_prop->multiProcessorCount;
+
+    const auto available_persistent_buffer_size =
+        sm_register_file_size * device_multiprocessor_count;
+
+    if (persistent_buffer_size > available_persistent_buffer_size) {
+      scheduler_debug_utils::canScheduleRejectReason(
+          ScheduleHeuristic::Persistent, "not enough registers for persistece");
```
Comment (csarofeen) [inline third_party/nvfuser/csrc/scheduler/registry.cpp]: Nit: persistence

File: third_party/nvfuser/csrc/scheduler/registry.cpp [hunk 9]
```diff
@@ -1839,6 +1842,202 @@ class PersistentKernelScheduler : public SchedulerEntry {
...
+      HeuristicSummary* data_cache,
+      const std::vector<TensorView*>& reduction_tvs,
+      const scheduler_utils::TvProperties& properties) {
+    FUSER_PERF_SCOPE("PersistentKernelScheduler::canScheduleRuntimeOuter");
+    FusionGuard fg(fusion);
+
+    const auto device_prop = at::cuda::getCurrentDeviceProperties();
+
+    const int64_t sm_register_file_size =
+        static_cast<int64_t>(device_prop->regsPerBlock * sizeof(int));
```
Comment (csarofeen) [inline third_party/nvfuser/csrc/scheduler/registry.cpp:1858]: Can't you preemptively subtract off the same values as in `getAvailableRegisterCount`?

File: third_party/nvfuser/csrc/scheduler/registry.cpp [hunk 10]
```diff
@@ -1839,6 +1842,202 @@ class PersistentKernelScheduler : public SchedulerEntry {
...
+    // doesn't seem to be noticeably bad on A100, though. For now,
+    // disable the schedule if not evenly divisible just on Titan RTX
+    // only.
+    // TODO: Revisit
+    if (is_cross_grid &&
+        (properties.total_iteration_numel %
+             (vectorization_factor * cross_grid_params->launch_params.bdimx() *
+              cross_grid_params->launch_params.gdimx()) !=
+         0) &&
+        device_prop->major == 7) {
```
Comment (csarofeen) [inline third_party/nvfuser/csrc/scheduler/registry.cpp:2034]: Isn't this disabling this on more than just titan rtx? Don't you need a minor version here?

File: third_party/nvfuser/test/test_gpu_outer_reduction.cpp [hunk 11]
```diff
@@ -1343,5 +1343,866 @@ TEST_F(
...
+
+// Forward welford
+void grid_persistent_welford_outer_norm_like_scheduler(
+    int64_t N,
+    int64_t HW,
+    int64_t C,
+    DataType dtype,
+    bool use_weights = false,
+    DataType weights_dtype = DataType::Float) {
+  const bool benchmark_mode = isBenchmarkMode();
```
Comment (csarofeen) [inline third_party/nvfuser/test/test_gpu_outer_reduction.cpp:1562]: Should there still be a benchmark mode in the tests? I assume this was for development but guessing it should probably be removed.

**Review by csarofeen — APPROVED** [review]
LGTM nice work.

File: third_party/nvfuser/csrc/scheduler/normalization_utils.h (hunk 4 above)
Comment (naoyam) [reply third_party/nvfuser/csrc/scheduler/normalization_utils.h]: Fixed to "Each of bdimx and gdimy determines bdym and gdimx, respecitively". bdimx and gdimy are determined as shown in the above loop nest, and bdimy and gdimx are picked accordingly.

File: third_party/nvfuser/csrc/scheduler/reduction_utils.cpp (hunk 5 above)
Comment (naoyam) [reply third_party/nvfuser/csrc/scheduler/reduction_utils.cpp:108]: Not exactly. The inner dimension of the split is not parallelized.

File: third_party/nvfuser/csrc/scheduler/reduction_utils.cpp (hunk 1 above)
Comment (naoyam) [reply third_party/nvfuser/csrc/scheduler/reduction_utils.cpp:261]: I'll work on this later as a separate PR.

File: third_party/nvfuser/csrc/scheduler/registry.cpp (hunk 9 above)
Comment (naoyam) [reply third_party/nvfuser/csrc/scheduler/registry.cpp:1858]: I think that should also work. Will consider in next iterations of PRs.

File: third_party/nvfuser/csrc/scheduler/registry.cpp (hunk 10 above)
Comment (naoyam) [reply third_party/nvfuser/csrc/scheduler/registry.cpp:2034]: Missed updating the last sentence of the comment above, but as mentioned in the first line, I had this issue with both Titan RTX and V100, so I just disabled anything with major version number of 7.

Comment (naoyam) [reply third_party/nvfuser/csrc/scheduler/registry.cpp:2034]: Clarified the comment.

File: third_party/nvfuser/test/test_gpu_outer_reduction.cpp (hunk 11 above)
Comment (naoyam) [reply third_party/nvfuser/test/test_gpu_outer_reduction.cpp:1562]: Let me keep it for now as there's still performance issues and having benchmarks here is very handy as both performance measurement and correctness validation can be done.

File: third_party/nvfuser/csrc/scheduler/reduction_utils.cpp (hunk 1 above)
Comment (naoyam) [reply third_party/nvfuser/csrc/scheduler/reduction_utils.cpp:261]: #2464
